### PR TITLE
[feature] Support complex matrix

### PR DESCRIPTION
**Pull request by soumyasen1809**
Related to issue #35 
- Added initial support for Complex Matrix

### CODE REVIEW

File: src/traits/mutable.rs
Comment (soumyasen1809) [inline src/traits/mutable.rs:22]: Note: In next PRs, this can be removed and the handling added to `pub trait MutMatrix` (with a generic)

File: src/traits/fp.rs
Comment (soumyasen1809) [inline src/traits/fp.rs:63]: Note: In next PRs, this can be removed and the handling added to `pub trait FPMatrix` (with a generic)

File: src/traits/fp.rs
Comment (soumyasen1809) [inline src/traits/fp.rs:96]: Note: In next PRs, this can be removed and the handling added to `pub trait FPMatrix` (with a generic)

File: src/complex/matrix.rs
Comment (Axect) [inline src/complex/matrix.rs:377]: Result shows still f64 matrix

File: src/complex/matrix.rs
Comment (Axect) [inline src/complex/matrix.rs:60]: Suggestion: I already write an alias for Complex64 as C64 in complex/mod.rs.
In example, how about use C64?

File: src/complex/matrix.rs
Comment (Axect) [inline src/complex/matrix.rs:1766]: Suggestion: Instead of commenting out unimplemented methods, consider using the `unimplemented!()` macro. This provides a clear indication that the functionality is planned but not yet implemented, and will panic if accidentally called. For example:

fn lu(&self) -> PQLU {
    unimplemented!("LU decomposition not yet implemented")
}

This approach maintains the method signatures in the trait, making it easier to track which methods still need implementation.

File: src/traits/fp.rs
Comment (Axect) [reply src/traits/fp.rs:96]: Regarding the use of Complex in `src/traits/{fp, mutable}.rs`:

1. Let's avoid using `use num_complex::Complex` directly in these files, as Complex should only be available with the "complex" feature enabled.

2. You mentioned that a generic implementation is planned for the future. Given this, I agree it would be best to start with the generic implementation now rather than implementing Complex-specific versions that will need to be modified later.

Let's pivot to implementing these traits generically from the outset. This approach will:
- Provide a more flexible foundation
- Reduce duplicate code when supporting both real and complex numbers
- Align with your planned future direction

Could you share your thoughts on how you'd like to approach this generic implementation? We can discuss any specific challenges or design considerations you foresee.

File: Cargo.toml
Comment (Axect) [inline Cargo.toml:35]: Regarding the matrixmultiply dependency in Cargo.toml:

The current specification includes the "cgemm" feature unconditionally:
matrixmultiply = { version = "0.3", features = ["threading", "cgemm"] }

However, the "cgemm" feature is already correctly tied to Peroxide's "complex" feature in the features section:
[features]
complex = [ ..., "matrixmultiply/cgemm"]

Suggestion:
Remove the "cgemm" feature from the direct dependency specification:
matrixmultiply = { version = "0.3", features = ["threading"] }

**Comment by Axect** [comment]
Hello, @soumyasen1809 

Thank you for your incredible work. I'm genuinely impressed by how comprehensively you've implemented almost everything needed for ComplexMatrix. It must have been an exhausting task, and I truly appreciate your effort.

I've reviewed the changes and left a few comments requesting some modifications. Please take a look at them when you have a chance. We can discuss any points further in the replies or you can proceed with the suggested changes as you see fit.

I noticed you've been working on the dev branch, which is great. However, I've created a new branch specifically for the Complex feature: 35-support-complex-number. It would be ideal if we could move this work to that branch to keep our feature development organized.

Once again, thank you for your hard work on this. Let's refine it together and get this fantastic addition to Peroxide across the finish line.

File: src/complex/matrix.rs
Comment (soumyasen1809) [reply src/complex/matrix.rs:1766]: In my opinion, I would rather remove the commented-out functions for now and add a ToDo note. The reason being, that to add the function signatures, the structs, enum etc needs to be defined/added. I am not a fan to adding such structs and enums before-hand and wait for it's use in the future.

For example,
```rust
fn lu(&self) -> PQLU {
unimplemented!("LU decomposition not yet implemented")
}
```
For this, I need to define PQLU in the code. There is no reason to define this now and wait for its usage sometime in the future. What do you think?

File: src/complex/matrix.rs
Comment (soumyasen1809) [reply src/complex/matrix.rs:60]: This is fine. But writing Complex64 makes it more obvious for the user. Either way it does not hurt.

**Comment by soumyasen1809** [comment]
Hi @Axect thank you for your kind words :)

> I've reviewed the changes and left a few comments requesting some modifications. Please take a look at them when you have a chance. We can discuss any points further in the replies or you can proceed with the suggested changes as you see fit.

Please find the requested changes in my next PR: https://github.com/Axect/Peroxide/pull/71/commits/f2778957a20f24e9ef709482a4a7e0013c1b7788
I have left 2 comments back to your comments (regarding C64 use and using unimplemented!).

> I noticed you've been working on the dev branch, which is great. However, I've created a new branch specifically for the Complex feature: 35-support-complex-number. It would be ideal if we could move this work to that branch to keep our feature development organized.

Also, I have changed the branch from `dev` to `35-support-complex`. In my opinion, this PR can be merged if you are okay with it. Other changes like using C64 etc. can be done on top of it later.
I have already started working on the remaining parts of the `ComplexMatrix` and will soon a add the next PR.

File: Cargo.toml
Comment (soumyasen1809) [reply Cargo.toml]: Done in PR: https://github.com/Axect/Peroxide/pull/71/commits/f2778957a20f24e9ef709482a4a7e0013c1b7788

File: src/complex/matrix.rs
Comment (soumyasen1809) [reply src/complex/matrix.rs:377]: Done in PR: https://github.com/Axect/Peroxide/pull/71/commits/f2778957a20f24e9ef709482a4a7e0013c1b7788

File: src/traits/mutable.rs
Comment (soumyasen1809) [reply src/traits/mutable.rs]: Done in PR: https://github.com/Axect/Peroxide/pull/71/commits/f2778957a20f24e9ef709482a4a7e0013c1b7788

File: src/traits/fp.rs
Comment (soumyasen1809) [reply src/traits/fp.rs]: Done in PR: https://github.com/Axect/Peroxide/pull/71/commits/f2778957a20f24e9ef709482a4a7e0013c1b7788

File: src/complex/matrix.rs
Comment (Axect) [reply src/complex/matrix.rs:1768]: I agree with your approach. It's a sensible way to keep the codebase clean and focused. Let's proceed with your suggestion.

**Comment by soumyasen1809** [comment]
Hi @Axect 
Once you have some time, please do review this PR. Once merged, I can rebase and add my next PRs.

**Comment by Axect** [comment]
Hello @soumyasen1809 

I apologize for the delay in reviewing your PR. I had gone through most of your code changes last week but held off on finalizing the review as I hadn't had the chance to test it locally.

After reviewing it today, I found that most of the changes look good. However, I noticed an error occurring in `tests/complex_matrix.rs`. Specifically, when running `cargo test --features complex`, it throws an error because `ComplexMatrix` and `Complex64` are being used without being declared.

To resolve this, we need to add the following import statements within the `test_seq` function:

```rust
use peroxide::complex::matrix::ComplexMatrix;
use num_complex::Complex64;
```

In the future, we might want to create a `fuga` for complex numbers to handle this more elegantly, but for now, if you could make these modifications, I'll be able to merge the PR right away.

Also, if you don't mind, it would be helpful to run both `cargo test` and `cargo test --features complex` before submitting PRs in the future. This can help catch potential errors early and ensure smoother integration.

Once you've made these changes, please let me know, and I'll proceed with the merge. After that, you can go ahead and rebase for your next PRs.

Thank you for your patience and contribution!

**Review by Axect — APPROVED** [review]

**Comment by soumyasen1809** [comment]
Hi @Axect I have fixed the issue in the next PR

> After reviewing it today, I found that most of the changes look good. However, I noticed an error occurring in `tests/complex_matrix.rs`. Specifically, when running `cargo test --features complex`, it throws an error because `ComplexMatrix` and `Complex64` are being used without being declared.
> 
> To resolve this, we need to add the following import statements within the `test_seq` function:
> 
> ```rust
> use peroxide::complex::matrix::ComplexMatrix;
> use num_complex::Complex64;
> ```
> 

Thanks for pointing that out.

> Also, if you don't mind, it would be helpful to run both `cargo test` and `cargo test --features complex` before submitting PRs in the future. This can help catch potential errors early and ensure smoother integration.

I forgot to run `cargo test --features complex`. My apologies for that. Extremely sorry!

**Pull request merged**
